File: Android/TUIKit/TUICallKit/tuicallkit/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/view/floatwindow/FloatWindowService.java
package com.tencent.qcloud.tuikit.tuicallkit.view.floatwindow;

import android.animation.ValueAnimator;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.PixelFormat;
import android.os.Binder;
import android.os.Build;
import android.os.IBinder;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;

import com.tencent.qcloud.tuikit.tuicallkit.R;

public class FloatWindowService extends Service {
    private static Intent        mStartIntent;
    private static FloatCallView mCallView;
    private static Context       mContext;

    private WindowManager              mWindowManager;
    private WindowManager.LayoutParams mWindowLayoutParams;

    private int mScreenWidth;
    private int mWidth;
    private int mTouchStartX;
    private int mTouchStartY;
    private int mTouchCurrentX;
    private int mTouchCurrentY;
    private int mStartX;
    private int mStartY;
    private int mStopX;
    private int mStopY;

    private boolean mIsMove;

    public static void startFloatService(Context context, FloatCallView callView) {
        mContext = context;
        mCallView = callView;
        mStartIntent = new Intent(context, FloatWindowService.class);
        context.startService(mStartIntent);
    }

    public static void stopService(Context context) {
        if (null != mStartIntent) {
            context.stopService(mStartIntent);
        }
    }

    @Override
    public void onCreate() {
        super.onCreate();
        initWindow();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return new FloatBinder();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (null != mCallView) {
            mWindowManager.removeView(mCallView);
            mCallView = null;
        }
    }

    private void initWindow() {
        mWindowManager = (WindowManager) getApplicationContext().getSystemService(Context.WINDOW_SERVICE);
        mWindowLayoutParams = getViewParams();
        mScreenWidth = mWindowManager.getDefaultDisplay().getWidth();

        if (null != mCallView) {
            mCallView.setBackgroundResource(R.drawable.tuicalling_bg_floatwindow_left);
            mWindowManager.addView(mCallView, mWindowLayoutParams);
            mCallView.setOnTouchListener(new FloatingListener());
        }
    }

    private WindowManager.LayoutParams getViewParams() {
        mWindowLayoutParams = new WindowManager.LayoutParams();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            mWindowLayoutParams.type = WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY;
        } else {
            mWindowLayoutParams.type = WindowManager.LayoutParams.TYPE_PHONE;
        }
        mWindowLayoutParams.flags = WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
                | WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                | WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS;

        mWindowLayoutParams.gravity = Gravity.LEFT | Gravity.TOP;
        mWindowLayoutParams.x = 0;
        mWindowLayoutParams.y = mWindowManager.getDefaultDisplay().getHeight() / 2;

        mWindowLayoutParams.width = WindowManager.LayoutParams.WRAP_CONTENT;
        mWindowLayoutParams.height = WindowManager.LayoutParams.WRAP_CONTENT;
        mWindowLayoutParams.format = PixelFormat.TRANSPARENT;

        return mWindowLayoutParams;
    }

    public class FloatBinder extends Binder {
        public FloatWindowService getService() {
            return FloatWindowService.this;
        }
    }

    private class FloatingListener implements View.OnTouchListener {
        @Override
        public boolean onTouch(View v, MotionEvent event) {
            int action = event.getAction();
            switch (action) {
                case MotionEvent.ACTION_DOWN:
                    mIsMove = false;
                    mTouchStartX = (int) event.getRawX();
                    mTouchStartY = (int) event.getRawY();

                    mStartX = (int) event.getRawX();
                    mStartY = (int) event.getRawY();
                    break;
                case MotionEvent.ACTION_MOVE:
                    mTouchCurrentX = (int) event.getRawX();
                    mTouchCurrentY = (int) event.getRawY();
                    if (mWindowLayoutParams != null && null != mCallView) {
                        mWindowLayoutParams.x += mTouchCurrentX - mTouchStartX;
                        mWindowLayoutParams.y += mTouchCurrentY - mTouchStartY;
                        mWindowManager.updateViewLayout(mCallView, mWindowLayoutParams);
                    }
                    mTouchStartX = mTouchCurrentX;
                    mTouchStartY = mTouchCurrentY;
                    break;
                case MotionEvent.ACTION_UP:
                    mStopX = (int) event.getRawX();
                    mStopY = (int) event.getRawY();
                    if (Math.abs(mStartX - mStopX) >= 5 || Math.abs(mStartY - mStopY) >= 5) {
                        mIsMove = true;
                        if (null != mCallView) {
                            mWidth = mCallView.getWidth();
                            if (mTouchCurrentX > (mScreenWidth / 2)) {
                                startScroll(mStopX, mScreenWidth - mWidth, false);
                            } else {
                                startScroll(mStopX, 0, true);
                            }
                        }
                    }
                    break;
                default:
                    break;
            }
            return mIsMove;
        }
    }

    private void startScroll(int start, int end, boolean isLeft) {
        ValueAnimator valueAnimator = ValueAnimator.ofFloat(start, end).setDuration(300);
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                if (mWindowLayoutParams == null || mCallView == null) {
                    return;
                }
                mWidth = mCallView.getWidth();
                if (isLeft) {
                    mWindowLayoutParams.x = (int) (start * (1 - animation.getAnimatedFraction()));
                    mCallView.setBackgroundResource(R.drawable.tuicalling_bg_floatwindow_left);
                } else {
                    float end = (mScreenWidth - start - mWidth) * animation.getAnimatedFraction();
                    mWindowLayoutParams.x = (int) (start + end);
                    mCallView.setBackgroundResource(R.drawable.tuicalling_bg_floatwindow_right);
                }
                calculateHeight();
                mWindowManager.updateViewLayout(mCallView, mWindowLayoutParams);
            }
        });
        valueAnimator.start();
    }

    private void calculateHeight() {
        if (mWindowLayoutParams == null) {
            return;
        }
        int height = mCallView.getHeight();
        int screenHeight = mWindowManager.getDefaultDisplay().getHeight();
        int resourceId = mContext.getResources().getIdentifier("status_bar_height",
                "dimen", "android");
        int statusBarHeight = mContext.getResources().getDimensionPixelSize(resourceId);
        if (mWindowLayoutParams.y < 0) {
            mWindowLayoutParams.y = 0;
        } else if (mWindowLayoutParams.y > (screenHeight - height - statusBarHeight)) {
            mWindowLayoutParams.y = screenHeight - height - statusBarHeight;
        }
    }
}

File: Android/TUIKit/TUICallKit/tuicallkit-kt/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/view/component/floatview/FloatWindowService.kt
package com.tencent.qcloud.tuikit.tuicallkit.view.component.floatview

import android.animation.ValueAnimator
import android.app.Service
import android.content.Intent
import android.content.res.Configuration
import android.graphics.PixelFormat
import android.os.Binder
import android.os.Build
import android.os.IBinder
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import com.tencent.qcloud.tuicore.ServiceInitializer
import com.tencent.qcloud.tuicore.util.ScreenUtil
import com.tencent.qcloud.tuikit.tuicallengine.TUICallDefine
import com.tencent.qcloud.tuikit.tuicallkit.state.TUICallState
import com.tencent.qcloud.tuikit.tuicallkit.view.CallKitActivity
import com.tencent.qcloud.tuikit.tuicallkit.view.root.BaseCallView

class FloatWindowService : Service() {
    private var windowManager: WindowManager? = null
    private var windowLayoutParams: WindowManager.LayoutParams? = null
    private var screenWidth = 0
    private var callViewWidth = 0
    private var touchStartX = 0
    private var touchStartY = 0
    private var touchCurrentX = 0
    private var touchCurrentY = 0
    private var startX = 0
    private var startY = 0
    private var stopX = 0
    private var stopY = 0
    private var isMove = false

    companion object {
        private var callView: BaseCallView? = null

        fun startFloatService(view: BaseCallView) {
            this.callView = view
            this.callView?.setOnClickListener {
                stopService()
                if (TUICallState.instance.selfUser.get().callStatus.get() != TUICallDefine.Status.None) {
                    val intent = Intent(ServiceInitializer.getAppContext(), CallKitActivity::class.java)
                    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                    ServiceInitializer.getAppContext().startActivity(intent)
                }
            }
            var serviceIntent = Intent(ServiceInitializer.getAppContext(), FloatWindowService::class.java)
            ServiceInitializer.getAppContext().startService(serviceIntent)
        }

        fun stopService() {
            if (callView != null) {
                callView?.clear()
            }
            var serviceIntent = Intent(ServiceInitializer.getAppContext(), FloatWindowService::class.java)
            ServiceInitializer.getAppContext().stopService(serviceIntent)
        }
    }

    override fun onCreate() {
        super.onCreate()
        initWindow()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (callView == null || windowLayoutParams == null || windowManager == null) {
            return
        }
        windowLayoutParams?.x = 0
        val screenWidth = ScreenUtil.getScreenWidth(applicationContext)
        val screenHeight = ScreenUtil.getScreenHeight(applicationContext)
        windowLayoutParams?.y = if (screenWidth > screenHeight) {
            (screenHeight - callView!!.height) / 2
        } else {
            ScreenUtil.dip2px(100f)
        }
        callView?.let { windowManager?.updateViewLayout(it, windowLayoutParams) }
    }

    override fun onBind(intent: Intent): IBinder {
        return FloatBinder()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        return START_NOT_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        if (null != callView && callView!!.isAttachedToWindow) {
            windowManager?.removeView(callView)
        }
        callView = null
    }

    private fun initWindow() {
        windowManager = applicationContext.getSystemService(WINDOW_SERVICE) as WindowManager
        windowLayoutParams = viewParams
        screenWidth = ScreenUtil.getScreenWidth(applicationContext)
        if (null != callView) {
            windowManager!!.addView(callView, windowLayoutParams)
            callView!!.setOnTouchListener(FloatingListener())
        }
    }

    private val viewParams: WindowManager.LayoutParams
        private get() {
            windowLayoutParams = WindowManager.LayoutParams()
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                windowLayoutParams!!.type = WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
            } else {
                windowLayoutParams!!.type = WindowManager.LayoutParams.TYPE_PHONE
            }
            windowLayoutParams!!.flags = (WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
                    or WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                    or WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
            windowLayoutParams!!.gravity = Gravity.END or Gravity.TOP
            windowLayoutParams!!.x = 10
            windowLayoutParams!!.y = ScreenUtil.dip2px(100f)
            windowLayoutParams!!.width = WindowManager.LayoutParams.WRAP_CONTENT
            windowLayoutParams!!.height = WindowManager.LayoutParams.WRAP_CONTENT
            windowLayoutParams!!.format = PixelFormat.TRANSPARENT
            return windowLayoutParams as WindowManager.LayoutParams
        }

    inner class FloatBinder : Binder() {
        val service: FloatWindowService
            get() = this@FloatWindowService
    }

    inner class FloatingListener : View.OnTouchListener {
        override fun onTouch(v: View, event: MotionEvent): Boolean {
            val action = event.action
            when (action) {
                MotionEvent.ACTION_DOWN -> {
                    isMove = false
                    touchStartX = event.rawX.toInt()
                    touchStartY = event.rawY.toInt()
                    startX = event.rawX.toInt()
                    startY = event.rawY.toInt()
                }

                MotionEvent.ACTION_MOVE -> {
                    touchCurrentX = event.rawX.toInt()
                    touchCurrentY = event.rawY.toInt()
                    if (windowLayoutParams != null && null != callView) {
                        windowLayoutParams!!.x += touchStartX - touchCurrentX
                        windowLayoutParams!!.y += touchCurrentY - touchStartY
                        windowManager!!.updateViewLayout(callView, windowLayoutParams)
                    }
                    touchStartX = touchCurrentX
                    touchStartY = touchCurrentY
                }

                MotionEvent.ACTION_UP -> {
                    stopX = event.rawX.toInt()
                    stopY = event.rawY.toInt()
                    if (Math.abs(startX - stopX) >= 5 || Math.abs(startY - stopY) >= 5) {
                        isMove = true
                        if (null != callView) {
                            callViewWidth = callView!!.width
                            if (touchCurrentX < screenWidth / 2) {
                                startScroll(screenWidth - callViewWidth, stopX, false)
                            } else {
                                startScroll(0, stopX, true)
                            }
                        }
                    }
                }

                else -> {}
            }
            return isMove
        }
    }

    private fun startScroll(start: Int, end: Int, isLeft: Boolean) {
        val valueAnimator = ValueAnimator.ofFloat(start.toFloat(), end.toFloat()).setDuration(300)
        valueAnimator.addUpdateListener(ValueAnimator.AnimatorUpdateListener { animation ->
            if (windowLayoutParams == null || callView == null) {
                return@AnimatorUpdateListener
            }
            callViewWidth = callView!!.width
            if (isLeft) {
                windowLayoutParams!!.x = (start * (1 - animation.animatedFraction)).toInt()
            } else {
                val end = (screenWidth - start - callViewWidth) * animation.animatedFraction
                windowLayoutParams!!.x = (start + end).toInt()
            }
            if (windowLayoutParams!!.x > screenWidth - callViewWidth) {
                windowLayoutParams!!.x = screenWidth - callViewWidth
            }
            calculateHeight()
            windowManager!!.updateViewLayout(callView, windowLayoutParams)
        })
        valueAnimator.start()
    }

    private fun calculateHeight() {
        if (windowLayoutParams == null) {
            return
        }
        val height = callView!!.height
        val screenHeight = ScreenUtil.getScreenHeight(applicationContext)
        val resourceId = ServiceInitializer.getAppContext().resources
            .getIdentifier("status_bar_height", "dimen", "android")
        val statusBarHeight = ServiceInitializer.getAppContext().resources.getDimensionPixelSize(resourceId)
        if (windowLayoutParams!!.y < 0) {
            windowLayoutParams!!.y = 0
        } else if (windowLayoutParams!!.y > screenHeight - height - statusBarHeight) {
            windowLayoutParams!!.y = screenHeight - height - statusBarHeight
        }
    }

}